File: spheretocube.go
package main

import (
	"os"
	"fmt"
	"image"
	"image/jpeg"
    "math"
    "time"
    "interp"
)

/**
 * 计算向量的长度
 * @param  {[type]} val [3]float64  [description]
 * @return {float64}     向量长度
 */
func length(val [3]float64) (float64){
    return math.Sqrt(val[0] * val[0] + val[1] * val[1] + val[2] * val[2])
}

/**
 * 向量除以常数计算
 * @param  {[type]} val [3]float64    [description]
 * @param  {[type]} d   float64       [description]
 * @return {[3]float64}
 */
func div(val [3]float64, d float64) ([3]float64) {
    return [3]float64{val[0] / d, val[1] / d, val[2] / d}
}

/**
 * 计算生成向量对应的单位向量
 * @param  {[type]} val [3]float64  [description]
 * @return {[3]float64}     单位向量
 */
func norm(val [3]float64) ([3]float64){
    return div(val, length(val))
}

/**
 * 向量相加计算
 * @param  {[type]} val [3]float64  [description]
 * @param  {[type]} vec [3]float64  [description]
 * @return {[3]float64}
 */
func add(val [3]float64, vec [3]float64) ([3]float64){
    return [3]float64{val[0] + vec[0], val[1] + vec[1], val[2] + vec[2]}
}

/**
 * 向量乘以数字计算
 * @param  {[type]} val [3]float64    [description]
 * @param  {[type]} d   float64       [description]
 * @return {float64}
 */
func mul(val [3]float64, d float64) ([3]float64) {
    return [3]float64{val[0] * d, val[1] * d, val[2] * d}
}

/**
 * 三位坐标转换成极坐标
 * @param {[type]} x float64  [description]
 * @param {[type]} y float64  [description]
 * @param {[type]} z float64) 
 * @return (hRad float64, vRad float64)
 */
func XYZToHVRad(x float64, y float64, z float64) (hRad float64, vRad float64){
    t := math.Sqrt( x * x + y * y)

    hRad = 0.0
    if t > 0.0 {
        cosh := x / t
        hRad = math.Acos(cosh)
        if y < 0.0 {
            hRad = 2.0 * math.Pi - hRad
        }
    }
    d := math.Sqrt(t*t + z*z)
    sint := z / d
    vRad = math.Asin(sint)

    return hRad, vRad
}

/**
 * 计算生成单张六面图的坐标映射数组
 * @param  {[type]} origin      [3]float64    [description]
 * @param  {[type]} wVec        [3]float64    [description]
 * @param  {[type]} hVec        [3]float64    [description]
 * @param  {[type]} imageWidth  int           [description]
 * @param  {[type]} imageHeight int           [description]
 * @param  {[type]} width       int           [description]
 * @param  {[type]} height      int)          (result       [][2]int [description]
 * @return {[type]}             [description]
 */
func calculatePosition(origin [3]float64, wVec [3]float64, hVec [3]float64, imageWidth int, imageHeight int, width int, height int) (result [][2]float64){
    wFace := length(wVec)
    hFace := length(hVec)
    wPixel := wFace / float64(width)
    hPixel := hFace / float64(height)

    wSubPixel := wPixel / 1.0
    hSubPixel := hPixel / 1.0

    wDir := norm(wVec)
    hDir := norm(hVec)

    var dy [3]float64
    var basePos [3]float64
    var dx [3]float64
    var pos [3]float64
    var dySubPixel [3]float64
    var baseSubPos [3]float64
    var dxSubPixel [3]float64
    var subPos [3]float64

    result = [][2]float64{}

    for py := 0; py < height; py++ {
        dy = mul(hDir, float64(py) * hPixel)
        basePos = add(origin, dy)
        for px := 0; px < width; px++ {
            dx = mul(wDir, float64(px) * wPixel)
            pos = add(basePos, dx)

            dySubPixel = mul(hDir, 0.5 * hSubPixel)
            baseSubPos = add(pos, dySubPixel)

            dxSubPixel = mul(wDir, 0.5 * wSubPixel)
            subPos = add(baseSubPos, dxSubPixel)

            hRad, vRad := XYZToHVRad(subPos[0], subPos[1], subPos[2])

            x := hRad * float64(imageWidth) / (math.Pi * 2.0)
            y := (vRad + math.Pi * 0.5) * float64(imageHeight) / math.Pi
            y = float64(imageHeight) - y

            result = append(result, [2]float64{x, y})
        }
    }

    return result
}

/**
 * 通过坐标映射数组以及原始图像，生成指定名称的六面图片
 * @param  {[type]} img       image.Image   图像对象
 * @param  {[type]} positions [][2]int      坐标映射数组
 * @param  {[type]} width     int           生成图片等宽
 * @param  {[type]} height    int           生成图片的高
 * @param  {[type]} path      string        生成图片保存位置
 * @return error
 */
func positionToImage(img image.Image, positions [][2]float64, width int, height int, path string) (err error) {
    rgba := image.NewRGBA(image.Rect(0, 0, width, height))
    for i := 0; i < len(positions); i++ {
        h := int(math.Floor(float64(i) / float64(width)))
        w := i - h * width
        color := interp.BilinearGeneral(img, positions[i][0], positions[i][1]);
        // color := img.At(positions[i][0], positions[i][1])
        rgba.Set(w, h, color)
    }
    newFile, _ := os.Create(path)
    defer newFile.Close()
    err = jpeg.Encode(newFile, rgba, &jpeg.Options{85})
    return err
}

// 入口主函数
func main() {
    layout := "01-02-2006 3.04.05 PM"
    fmt.Println(time.Now().Format(layout))

	file := "pano1.jpg"
	fimigfile, _ := os.Open(file)
    defer fimigfile.Close()
    img, _, err := image.Decode(fimigfile)
    if err != nil {
    	fmt.Printf("Decode image failed.")
    }else{
        bound := img.Bounds()
        imageWidth := bound.Max.X - bound.Min.X
        imageHeight := bound.Max.Y - bound.Min.Y
    	width := int(math.Ceil(float64(imageWidth) / math.Pi))
    	height := width
        fmt.Println(width)

        front := calculatePosition([3]float64{-1.0, 1.0, 1.0}, [3]float64{0.0, -2.0, 0.0}, [3]float64{0.0, 0.0, -2.0}, imageWidth, imageHeight, width, height)
        err = positionToImage(img, front, width, height, "pano_f.jpg")

        left := calculatePosition([3]float64{1.0, 1.0, 1.0}, [3]float64{-2.0, 0.0, 0.0}, [3]float64{0.0, 0.0, -2.0}, imageWidth, imageHeight, width, height)
        err = positionToImage(img, left, width, height, "pano_l.jpg")

        right := calculatePosition([3]float64{-1.0, -1.0, 1.0}, [3]float64{2.0, 0.0, 0.0}, [3]float64{0.0, 0.0, -2.0}, imageWidth, imageHeight, width, height)
        err = positionToImage(img, right, width, height, "pano_r.jpg")

        back := calculatePosition([3]float64{1.0, -1.0, 1.0}, [3]float64{0.0, 2.0, 0.0}, [3]float64{0.0, 0.0, -2.0}, imageWidth, imageHeight, width, height)
        err = positionToImage(img, back, width, height, "pano_b.jpg")

        up := calculatePosition([3]float64{1.0, 1.0, 1.0}, [3]float64{0.0, -2.0, 0.0}, [3]float64{2.0, 0.0, 0.0}, imageWidth, imageHeight, width, height)
        err = positionToImage(img, up, width, height, "pano_u.jpg")

        down := calculatePosition([3]float64{-1.0, 1.0, -1.0}, [3]float64{0.0, -2.0, 0.0}, [3]float64{2.0, 0.0, 0.0}, imageWidth, imageHeight, width, height)
        err = positionToImage(img, down, width, height, "pano_d.jpg")
    }

    fmt.Println(time.Now().Format(layout))
}

File: spheretocube.py
# coding=utf-8
#import Image
import os, math, time, multiprocessing
from PIL import Image

class Vector3f:
    def __init__(self):
        self.x = 0.0
        self.y = 0.0
        self.z = 0.0
        return

    def set(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z
        return

    def add(self, o):
        r = Vector3f()
        r.x = self.x + o.x
        r.y = self.y + o.y
        r.z = self.z + o.z
        return r

    def sub(self, o):
        r = Vector3f()
        r.x = self.x - o.x
        r.y = self.y - o.y
        r.z = self.z - o.z
        return r

    def mul(self, v):
        r = Vector3f()
        r.x = self.x * v
        r.y = self.y * v
        r.z = self.z * v
        return r

    def div(self, v):
        r = Vector3f()
        r.x = self.x / v
        r.y = self.y / v
        r.z = self.z / v
        return r

    def dot(self, o):
        return self.x * o.x + self.y * o.y + self.z * o.z

    def length(self):
        return math.sqrt(self.x * self.x + self.y * self.y + self.z * self.z)

    def norm(self):
        d = self.length()
        r = Vector3f()
        r.x = self.x / d
        r.y = self.y / d
        r.z = self.z / d
        return r

    pass

class UnitSphere:

    def __init__(self, texPath = None):
        self.TexturePath = ''
        self.PixelBuffer = None
        self.ImageW = 0
        self.ImageH = 0
        self.pi2 = 2 * math.pi
        self.pihalf = 0.5 * math.pi
        self.R = 0
        self.Hr = 0
        if texPath is not None:
            self.LoadTexture(texPath)
        return

    def LoadTexture(self, texPath):
        if not os.path.exists(texPath):
            raise

        print("loading %s" %(texPath))
        currentTime = time.time()
        self.TexturePath = texPath
        texImg = Image.open(texPath)

        # Memory will be used in large numbers, when use this code.
        # d = texImg.getdata()
        # self.PixelBuffer = list(d)
        
        self.PixelBuffer = texImg.load()
        self.ImageW = texImg.size[0]
        self.ImageH = texImg.size[1]
        # 半径
        self.R = self.ImageW / self.pi2
        # 高的半径，默认为二比一的话，那么就是半径的一半
        self.Hr = self.ImageH / math.pi

        print('Use time ' + str(time.time() - currentTime) + 'sec')

        return

    def GetColor(self, hRad, vRad):
        if self.PixelBuffer is None: return 0, 0, 0

        x = hRad * self.R
        y = (vRad + self.pihalf) * self.Hr
        y = self.ImageH - y

        px = int(math.floor(x))
        py = int(math.floor(y))

        return self.PixelBuffer[px, py]

        # idx = py * self.ImageW + px
        # return self.PixelBuffer[idx]

    pass

def XYZToHVRad(x, y, z):
    t = math.sqrt(x*x + y*y)

    hRad = 0.0
    if t > 0.000001: # t == 0.0
        cosh = x / t
        hRad = math.acos(cosh)
        if y < 0.0:
            hRad = 2.0 * math.pi - hRad

    d = math.sqrt(t*t + z*z)
    sint = z / d
    vRad = math.asin(sint)

    return hRad, vRad

# sampleRate采样范围，为了颜色值过度更加平滑
def RayCastCubeFace(sphere, origin, wVec, hVec, pxWidth, pxHeight, sampleRate):
    # input face rect, sampling rate
    # for each pixel, cast ray, and merge samples
    # save the face
    samples = float(sampleRate * sampleRate)

    wFace = wVec.length()
    hFace = hVec.length()

    wPixel = wFace / float(pxWidth)
    hPixel = hFace / float(pxHeight)
    wSubPixel = wPixel / sampleRate
    hSubPixel = hPixel / sampleRate

    wDir = wVec.norm()
    hDir = hVec.norm()

    print('RayCastCubeFace')
    currentTime = time.time()

    colors = []
    for py in range(pxHeight):
        dy = hDir.mul(py * hPixel)
        basePos = origin.add(dy)
        for px in range(pxWidth):
            dx = wDir.mul(px * wPixel)
            pos = basePos.add(dx)
            #c = RayCastPixel(pos, wPixel, hPixel, sampleRate)

            c = [0, 0, 0]
            for j in range(sampleRate):
                dySubPixel = hDir.mul((j + 0.5) * hSubPixel)
                baseSubPos = pos.add(dySubPixel)
                for i in range(sampleRate):
                    dxSubPixel = wDir.mul((i + 0.5) * wSubPixel)
                    subPos = baseSubPos.add(dxSubPixel)

                    hRad, vRad = XYZToHVRad(subPos.x, subPos.y, subPos.z)
                    r, g, b = sphere.GetColor(hRad, vRad)
                    c[0] += r
                    c[1] += g
                    c[2] += b
                    pass
            # end iterating subpixel

            r = round(c[0] / samples)
            g = round(c[1] / samples)
            b = round(c[2] / samples)
            colors.append((int(r), int(g), int(b)))
            pass
        pass
    # end iterating all pixels
    print('RayCastCubeFace time ' + str(time.time() - currentTime) + 'sec')
    return colors

def SaveCubeFace(path, w, h, colors):
    img = Image.new('RGB', (w, h))
    img.putdata(colors)
    img.save(path, quality=85)
    return

class RectDesc:

    def __init__(self):
        self.origin = Vector3f()
        self.hside = Vector3f()
        self.vside = Vector3f()
        return

    pass

def GetCubeFrontRect():
    desc = RectDesc()
    desc.origin.set(-1.0, 1.0, 1.0)
    desc.hside.set(0.0, -2.0, 0.0)
    desc.vside.set(0.0, 0.0, -2.0)
    return desc

def GetCubeRightRect():
    desc = RectDesc()
    desc.origin.set(-1.0, -1.0, 1.0)
    desc.hside.set(2.0, 0.0, 0.0)
    desc.vside.set(0.0, 0.0, -2.0)
    return desc

def GetCubeLeftRect():
    desc = RectDesc()
    desc.origin.set(1.0, 1.0, 1.0)
    desc.hside.set(-2.0, 0.0, 0.0)
    desc.vside.set(0.0, 0.0, -2.0)
    return desc

def GetCubeBackRect():
    desc = RectDesc()
    desc.origin.set(1.0, -1.0, 1.0)
    desc.hside.set(0.0, 2.0, 0.0)
    desc.vside.set(0.0, 0.0, -2.0)
    return desc

def GetCubeTopRect():
    desc = RectDesc()
    desc.origin.set(1.0, 1.0, 1.0)
    desc.hside.set(0.0, -2.0, 0.0)
    desc.vside.set(-2.0, 0.0, 0.0)
    return desc

def GetCubeBottomRect():
    desc = RectDesc()
    desc.origin.set(-1.0, 1.0, -1.0)
    desc.hside.set(0.0, -2.0, 0.0)
    desc.vside.set(2.0, 0.0, 0.0)
    return desc

# 多线程
def createCubeFront(origin, file, width, height, sample):
    global sphere
    print('generating front face')
    face = GetCubeFrontRect()
    colors = RayCastCubeFace(sphere, face.origin, face.hside, face.vside, width, height, sample)
    # print('save front face')
    SaveCubeFace(file, width, height, colors)
    print('end front face')
    return

def createCubeLeft(origin, file, width, height, sample):
    global sphere
    print('generating right face')
    face = GetCubeLeftRect()
    colors = RayCastCubeFace(sphere, face.origin, face.hside, face.vside, width, height, sample)
    # print('save left face')
    SaveCubeFace(file, width, height, colors)
    print('end left face')
    return

def createCubeRight(origin, file, width, height, sample):
    global sphere
    print('generating left face')
    face = GetCubeRightRect()
    colors = RayCastCubeFace(sphere, face.origin, face.hside, face.vside, width, height, sample)
    # print('save right face')
    SaveCubeFace(file, width, height, colors)
    print('end right face')
    return

def createCubeBack(origin, file, width, height, sample):
    global sphere
    print('generating back face')
    face = GetCubeBackRect()
    colors = RayCastCubeFace(sphere, face.origin, face.hside, face.vside, width, height, sample)
    # print('save back face')
    SaveCubeFace(file, width, height, colors)
    print('end back face')
    return

def createCubeTop(origin, file, width, height, sample):
    global sphere
    print('generating top face')
    face = GetCubeTopRect()
    colors = RayCastCubeFace(sphere, face.origin, face.hside, face.vside, width, height, sample)
    # print('save top face')
    SaveCubeFace(file, width, height, colors)
    print('end top face')
    return

def createCubeBottom(origin, file, width, height, sample):
    global sphere
    print('generating bottom face')
    face = GetCubeBottomRect()
    colors = RayCastCubeFace(sphere, face.origin, face.hside, face.vside, width, height, sample)
    # print('save bottom face')
    SaveCubeFace(file, width, height, colors)
    print('end bottom face')
    return

# 多进程处理全景
def run_process(file, w, h, s, num):
    # input unit cube and sphere
    # work on each cube face

    # 启动多进程
    pool = multiprocessing.Pool(num)
    pool.apply_async(createCubeFront, (file, 'f.jpg', w, h, s))
    pool.apply_async(createCubeLeft, (file, 'l.jpg', w, h, s))
    pool.apply_async(createCubeRight, (file, 'r.jpg', w, h, s))
    pool.apply_async(createCubeBack, (file, 'b.jpg', w, h, s))
    pool.apply_async(createCubeTop, (file, 'u.jpg', w, h, s))
    pool.apply_async(createCubeBottom, (file, 'd.jpg', w, h, s))
    pool.close()
    pool.join()
    return 

# if __name__ == '__main__':
#     sphere = UnitSphere()
#     sphere.LoadTexture('pano1.jpg')
#     createCubeFront('pano1.jpg', 'f.jpg', 1500, 1500, 1)

sphere = UnitSphere()
sphere.LoadTexture('pano1.jpg')

if __name__ == '__main__':
    print('go main')
    multiprocessing.freeze_support()
    print('main')
    # 主进程不处理数据
    currentTime = time.time()
    cpu_count = multiprocessing.cpu_count()
    print("The number of CPU is: " + str(cpu_count))
    process_num = cpu_count - 1 if cpu_count > 1 else 1
    print("The number of process is: " + str(process_num))
    run_process('pano1.jpg', 1592, 1592, 1, process_num)
    print('Use time ' + str(time.time() - currentTime) + 'sec')
# else:
#     sphere.LoadTexture('pano1.jpg')